### PR TITLE
Improved neovim floating windows, and misc. other improvements

### PR DESCRIPTION
**Pull request by subnut**
Please see the commit messages for details

EDIT:
My main goal in this PR is to improve the (currently deplorable) state of the completion-documentation popup provided by `vim-lsp`. It just so happened that I began by improving the `lsp#ui#vim#.....()` functions and ended up with a re-write of `documentation.vim`, so I simply included both :smile: 

As a (good) side-effect of making the `lsp#vim#ui` functions relative to the global-grid for neovim, now all popup are accurately positioned, and their positioning does not depend on the current window, splits, tab-line, cmdheight (which is variable in neovim, BTW. If you want, you can make it occupy the whole screen :grimacing:) etc.

### CODE REVIEW

**Comment by subnut** [comment]
@prabirshrestha Please check the failing macOS workflow. The error seems to be unrelated to this PR.

**Comment by thomasfaingnaert** [comment]
@subnut The failing workflow is just because the latest Neovim nightly build has no artifacts for Mac OS yet, so you can safely ignore it.

**Comment by subnut** [comment]
@thomasfaingnaert Thanks! :smile: 

So.... will this be merged by you? Or are we going to wait for other collaborators' opinions?

**Comment by thomasfaingnaert** [comment]
It's a bit busy at the moment, so I'd rather have @prabirshrestha, @hrsh7th, or @matn look at this.

**Comment by subnut** [comment]
@thomasfaingnaert I think you mean @mattn instead of @matn :wink:

**Comment by thomasfaingnaert** [comment]
Oops, fixed :smile:

**Comment by subnut** [comment]
New issue discovered. 
The popup menu gets frozen. Fix available with me. Implementing soon.

Please see the GIF below
gifcast_201023103412

**Comment by subnut** [comment]
Fixed! See GIF -
gifcast_201023121033

**Comment by subnut** [comment]
@prabirshrestha, @hrsh7th, @mattn , @thomasfaingnaert 
Please review and test. I tested all possible cases I could think of.

**Comment by prabirshrestha** [comment]
@subnut thanks for the PR I will have a look at it during the weekend. Popup has been one of the things i have wanted to fix for quite some time so good to see traction here.

I have also been working on and off on the popup api in vital.vim which you can find at https://github.com/vim-jp/vital.vim/pull/748. there are still some missing feature but already contains majority of the feature. goal is to ship the `popup.vim` as part of vim-lsp and also other plugins can take advantage of it. There are also few other things I would like to add such as margin/border/padding/title and so on.

You can also find some of my thoughts at https://github.com/vim-jp/vital.vim/issues/747.

as for `<silent>` free free to send another PR.

**Comment by subnut** [comment]
@prabirshrestha Yes, I'd seen that PR before. But I think that till that PR gets merged and is eventually integrated in vim-lsp in the future, we should improve the current state of the floating window, right?

As it stands right now (before this PR) the popup (**especially** the documentation popup, i.e. the on that is triggered on CompleteChanged) is unusable in neovim if you use splits. The vim popups are reasonably accurate/perfect. So, why not the same for neovim?

I think that the correct course of action would be to merge this PR for now, continue working on that API. When in the (near?) future that API is finalized, throw away all the current wrappers and use the new API. :partying_face: 

I believe that there should be one single wrapper/API that is used by all other plugins, so that development can be concentrated in one place. But I think that till that future API arrives, we should improve what we currently have in our hands.

Also, I the `<silent>` mappings are relevant to _this_ PR because the default algorithm doesn't use `screenrow()` or `screencol()`, so the `<silent>` is irrelevant

**Comment by subnut** [comment]
@prabirshrestha If you are worried about having to test my algorithm, then rest assured. I have tested it extensively myself. In fact I have spent 3 whole days on testing and perfecting before submitting this PR, and even after submitting, still continue to test and correct any errors in my part (see all those forced updates? :grin: )

I am going to extensively use this plugin for the coming 4 months, so rest assured, I will kill all bugs :beetle:

**Comment by prabirshrestha** [comment]
we don't have to wait for complete rewrite with vital.vim. This PR should be able to be merged after review. I will spend sometime during the weekend to play with it a bit more. i dont expect the vital.vim Pr to be merged there anytime soon. We can use popup.vim from as part of vim-lsp even before it gets merged but this is a bigger refactor and can be done separately.

In the meantime here is a bug I saw. This exists even before this PR.

lastest vim on windows.

image

neovim 0.4 on windows.

image

neovim seems to be single line instead of multiple lines.

**Review by prabirshrestha — CHANGES_REQUESTED** [review]

**Comment by subnut** [comment]
> In the meantime here is a bug I saw. neovim seems to be single line instead of multiple lines.

Well I tried to implement a workaround for it, but failed. I'll see what I can do over the next week....

File: plugin/lsp.vim
Comment (subnut) [reply plugin/lsp.vim:25]: Ok, will do..

Comment (subnut) [reply plugin/lsp.vim:25]: Done. Please check.

**Comment by prabirshrestha** [comment]
Got another bug in neovim nightly on linux. NVIM v0.5.0-756-g288f7f855

image

You will need to use `float2nr` if you are doing any divisions.

**Comment by subnut** [comment]
Oops.. OK, will look into this tommorrow..
But could you please check if the documentation is correct? (Last commit)

**Comment by subnut** [comment]
@prabirshrestha Looking closer to your error message, it says that the error occurred at line 39 of `s:show_documentation()`, which starts at line 13, so that brings us to line 52, which is -
```vim
let l:height = max([&lines - &cmdheight - l:row, &previewheight])
```

Now, previously I didn't know that the `max()` and `min()` functions do not work with floats (:sweat_smile:)
But look carefully, the arguments of the `max()` are `&lines - &cmdheight - l:row` and `&previewheight`
We know that `&lines`,`&cmdheight` and `&previewheight` cannot be floats (try setting them to floats), so that leaves us with `l:row`

But `l:row` can only have either `0` or `&lines`, (see line 42 and 46), as those are the only place where it is defined before the `max()` at line 52. But neither of `0` and `&lines` can be floats. So, `l:row` also cannot be a float !

Also, `integer - integer` gives us `integer`.
If all arguments are integers, then why did the `max()` freak out?

Seriously confused.

**Comment by prabirshrestha** [comment]
I debuged a bit further and seems like `row` and `col` for `v:event` in neovim is a float. Had to add manually convert it and it worked.

```vim
let a:event['row'] = float2nr(a:event['row'])
let a:event['col'] = float2nr(a:event['col'])
```

I'm running neovim nightly so not sure if it is a regression in neovim. Value was something like `12.0`.

Are you running neovim 0.4?

**Comment by subnut** [comment]
> Are you running neovim 0.4?

Yep!

**Comment by subnut** [comment]
@prabirshrestha Please check if the latest commit fixes that issue.

**Comment by prabirshrestha** [comment]
would be good if you could also file a bug in neovim. It would be good if `float2nr` was fixed in `s:complete_done()` so that `s:show_documentation` can use `a:event`. hopefully if neovim fixes it before their 0.5 release we could then remove this hack if 0.4 doesn't have this bug.

**Comment by subnut** [comment]
Hm... Ok. I will open an issue in neovim, but as I don't have first-hand experience regarding this issue, you will have to mention the details there.

Or maybe, in my opinion, since you experienced this issue first-hand, you should file the issue, don't you think?

**Comment by subnut** [comment]
@prabirshrestha I see this PR hasn't been merged yet. Any bugs?

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:32]: Please fix indentations.

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:61]: omit empty line here

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:98]: ditto

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:107]: please fix indentations above.

File: plugin/lsp.vim
Comment (mattn) [inline plugin/lsp.vim:114]: `:execute` is not a function. This is invalid usage of `:execute`.

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:127]: This two variables are not used in following.

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:93]: Please fix indentation here.

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:66]: omit empty line here

File: autoload/lsp/ui/vim/documentation.vim
Comment (mattn) [inline autoload/lsp/ui/vim/documentation.vim:53]: omit empty line here

**Comment by mattn** [comment]
Please rebase this pull-request from master. Thank you.

**Comment by subnut** [comment]
@mattn Please check whether the changes you requested have been implemented, or did I miss something.
<br>

> Please rebase this pull-request from master

Please tell me how to do that. I don't know actually, I am a newbie in git :sweat_smile:

**Comment by subnut** [comment]
@mattn This is the first time I have ever done a `rebase`
Did I do it correctly?

**Comment by subnut** [comment]
Ping @mattn @prabirshrestha

File: autoload/lsp/ui/vim/documentation.vim
Comment (hrsh7th) [inline autoload/lsp/ui/vim/documentation.vim:6]: It should be `complete_changed`.

File: autoload/lsp/ui/vim/documentation.vim
Comment (hrsh7th) [inline autoload/lsp/ui/vim/documentation.vim:10]: It should be debounced.

```viml
let s:last_timer_id = 0

function! s:complete_done() abort
    if !g:lsp_documentation_float | return | endif
    " Use a timer to avoid textlock (see :h textlock).
    let l:event = copy(v:event)
    call timer_stop(s:last_timer_id)
    let s:last_timer_id = timer_start(80, {-> s:show_documentation(l:event)})
endfunction
```

**Comment by hrsh7th** [comment]
Thank you for your contribution.

I checked the behaviors on neovim, it works well.

**Review by mattn — CHANGES_REQUESTED** [review]

File: autoload/lsp/ui/vim/documentation.vim
Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim:6]: I didn't change it!
It was `s:complete_done()` before the PR, and I don't want to change it unless @prabirshrestha says so....

Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim:6]: @prabirshrestha ?

File: autoload/lsp/ui/vim/documentation.vim
Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim:10]: Ok... but why?

Comment (hrsh7th) [reply autoload/lsp/ui/vim/documentation.vim:10]: If we keep pressing `<C-n>` to choose item at the bottom of pum, the current implementation will send requests for each all items. It bit bad for performance.

Sorry. I know you don't touch it on this PR but it's good chance to improve it, I thought.

File: plugin/lsp.vim
Comment (subnut) [reply plugin/lsp.vim]: `<silent>` is needed because without it the `screenrow()` and `screencol()` functions in neovim always return `1`, and shall result in incorrect positioning of floating windows.

As I am not sure which of these functions create floating windows, I used `<silent>` for all of them
<br>
We need to have it here because we cannot expect the user to use <silent> in their mappings.

Comment (subnut) [reply plugin/lsp.vim]: As for the `execute()`, I cannot recall why I added them, so I have removed them in the latest force-push.

File: autoload/lsp/ui/vim/documentation.vim
Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim:16]: Implemented. Please check and test.

**Comment by subnut** [comment]
@hrsh7th Please check my implementation of debouncing

File: plugin/lsp.vim
Comment (mattn) [inline plugin/lsp.vim:116]: Please remove `<silent>`

Comment (subnut) [reply plugin/lsp.vim:116]: On this line only?

Comment (subnut) [reply plugin/lsp.vim:116]: @mattn If you would help me identify which of these mappings __do not__ show popup, I will remove `<silent>` from them.
<br>
To see why `<silent>` __must__ be used with mappings that produce popup, see https://github.com/prabirshrestha/vim-lsp/pull/921#discussion_r521479419

**Comment by subnut** [comment]
@mattn If you would help me identify which of these mappings __do not__ show popup, I will remove `<silent>` from them.
<br>
To see why `<silent>` __must__ be used with mappings that produce popup, see https://github.com/prabirshrestha/vim-lsp/pull/921#discussion_r521479419

File: plugin/lsp.vim
Comment (subnut) [reply plugin/lsp.vim]: Done!

**Comment by subnut** [comment]
@hrsh7th I removed the `<silent>` from the mappings. I had originally added them because otherwise the neovim floating windows malfunctioned. But now, it seems the don't malfunction even if `<silent>` is not added.

Could you please test?

File: autoload/lsp/ui/vim/documentation.vim
Comment (hrsh7th) [inline autoload/lsp/ui/vim/documentation.vim:133]: I think we should add `call timer_stop(s:last_timer_id)` in here.
Because `s: show_documentation` is now debounced so I afraid to open popup after call` s: close_popup`.

Comment (hrsh7th) [reply autoload/lsp/ui/vim/documentation.vim:133]: Other debounce related codes are looks good to me.

**Comment by hrsh7th** [comment]
> @hrsh7th I removed the <silent> from the mappings. I had originally added them because otherwise the neovim floating windows malfunctioned. But now, it seems they don't malfunction even if <silent> is not added.
> 
>Could you please test?

OK. I will test it.

File: autoload/lsp/ui/vim/documentation.vim
Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim:132]: I added it in the latest commit. Please check if correctly implemented.

File: plugin/lsp.vim
Comment (hrsh7th) [inline plugin/lsp.vim:24]: It should be 100 by default.

I think this debouncing has no bad effect so we should enable debouncing by default.

I checked the current debouncing behaviors that worked well.

Comment (subnut) [reply plugin/lsp.vim:24]: OK

Comment (subnut) [reply plugin/lsp.vim:24]: I think it should be kept disabled by default (i.e. value should be 0)... I
 want to hear what other say.

<br>

@mattn @prabirshrestha @thomasfaingnaert What do you think? What should be the default value?

File: autoload/lsp/ui/vim/output.vim
Comment (hrsh7th) [inline autoload/lsp/ui/vim/output.vim:118]: I don't know the reason but this and L159 changes make exception sometimes... Hm...

Comment (hrsh7th) [reply autoload/lsp/ui/vim/output.vim:118]: My repro steps are the below.


0. Add `let g:lsp_diagnostics_float_cursor = 1`
1. Open some `go` file
2. Edit `fmt.Println()` (this should be marked as error by diagnostics)
3. Put cursor on `P` and wait for display floating diagnostic window.
4. Do `LspHover`.

Comment (subnut) [reply autoload/lsp/ui/vim/output.vim:118]: You mean this - (GIF below)
gifcast_201113113115

<br>

Seems OK to me...
Sorry if I did some step wrong.. I have never coded Go

Comment (subnut) [reply autoload/lsp/ui/vim/output.vim:118]: Are you using Neovim nightly?

Comment (hrsh7th) [reply autoload/lsp/ui/vim/output.vim:118]: I took screenshot please check it.

Kapture 2020-11-13 at 15 17 16

Comment (hrsh7th) [reply autoload/lsp/ui/vim/output.vim:118]: Yes, I'm using neovim nightly.

Comment (subnut) [reply autoload/lsp/ui/vim/output.vim:118]: I tried to fix it in the latest commit. Please check.

Comment (hrsh7th) [reply autoload/lsp/ui/vim/output.vim:118]: Thank you! It works perfectly.

**Comment by hrsh7th** [comment]
I think it almost ready to merge.

I think the remaining consideration points are `debouncing by default or not`.

File: autoload/lsp/ui/vim/documentation.vim
Comment (hrsh7th) [reply autoload/lsp/ui/vim/documentation.vim:7]: It is a trivial thing but I think `complete_changed` is better name.
See https://github.com/prabirshrestha/vim-lsp/pull/921/files#diff-b569a8722a789fc65d0920e66f13bfb4282225eb0b8260e254b72f073cca4828R149

Comment (subnut) [reply autoload/lsp/ui/vim/documentation.vim]: OK. Renamed.

File: plugin/lsp.vim
Comment (prabirshrestha) [reply plugin/lsp.vim:24]: for perf it is better to debounce primarily since vim script is very slow.

Comment (subnut) [reply plugin/lsp.vim:24]: OK.. so debounce by default.

Comment (prabirshrestha) [reply plugin/lsp.vim:24]: i dont have a strong opinion on this. can we be easily changed later as long as it is document on what it does and what values it should be wouldnt matter.

Comment (subnut) [reply plugin/lsp.vim:24]: OK. So I am setting the default to 80ms as suggested by @hrsh7th

**Comment by subnut** [comment]
@prabirshrestha Is the documentation of `g:lsp_documentation_debounce` OK?

**Comment by subnut** [comment]
> I think it almost ready to merge.
> 
> I think the remaining consideration points are debouncing by default or not.

@hrsh7th 
The remaining consideration point has been resolved
Now ready to merge?

**Review by hrsh7th — APPROVED** [review]
I think it ok to merge.

**Review by prabirshrestha — APPROVED** [review]

**Comment by subnut** [comment]
@prabirshrestha Who will merge this?

**Comment by hrsh7th** [comment]
I will if all requested changes are resolved.

**Comment by subnut** [comment]
@hrsh7th All requested changes have already been resolved, IIRC
If not, please show me which one hasn't been resolved. I will resolve it immediately.

**Comment by subnut** [comment]
Are we waiting for approval of @mattn ?

**Comment by subnut** [comment]
I repeat, *Are there __any__ unresolved requested changes?* If yes, **please show me which one.**

**Comment by hrsh7th** [comment]
I mean the PR still has request changes status.

But I think it OK to merge (I checked all inline comments already resolved)

**Comment by prabirshrestha** [comment]
Thanks everyone. Great change.

**Comment by mattn** [comment]
I don't make sure that it is related on this PR but current implementation show documentation of candidates in popup in broken location.

image

**Comment by subnut** [comment]
@mattn I am sure this is un-related, because I did not touch the completion _at all_ 
I only touched popups, _documentation_ popup.

Also, if you are on vim, then there should not have been any change at all. This PR only changed neovim behaviour.

**Comment by subnut** [comment]
@mattn Did you manage to fix the above issue? If yes, how,?

**Comment by mattn** [comment]
Sorry, it was already fixed. https://github.com/prabirshrestha/vim-lsp/commit/a56304f238f5368ad6c1b3b14b96308edd25a1f9

**Pull request merged**
